### ISSUE TITLE
License statement not accepted

### ISSUE THREAD

**Issue by netravnen** [OpenUserJS/OpenUserJS.org#1314]
My license statement in my script reads: `// @license         CC BY-SA 3.0; http://creativecommons.org/licenses/by-sa/3.0/` (one-line) ([GitHub][2], [OpenUserJS][1])

But the website returns me when I try to update my script on the page manually by doing 'Edit Script' > 'Submit Source Code' (Because the (github->openuserjs) webhook does not work for the same reason, I presume)
<img width="342" alt="2017-12-30 19_42_45-400 _ license is not osi primary and compatible in the metadata block s _ o" src="https://user-images.githubusercontent.com/1938389/34456557-ac8eb2f6-ed99-11e7-9477-6bcaae27aa66.png">

Can a/the reason be the 'CC BY-XYZ' licenses is not available fx here, in this dropdown ? :
<img width="629" alt="2017-12-30 19_38_09-edit torrentz2 dot eu magnet link _ openuserjs" src="https://user-images.githubusercontent.com/1938389/34456587-60ca8010-ed9a-11e7-87da-9d139acd1ea4.png">

Would be nice to have it fixed. So amongst other the license in the info section is not displayed wrongly like this.
<img width="368" alt="2017-12-30 19_50_13-about _ torrentz2 dot eu magnet link _ userscripts _ openuserjs" src="https://user-images.githubusercontent.com/1938389/34456604-ca9b3ce6-ed9a-11e7-8b64-f6c0fb114eba.png">

NB: Have tried looking around the OpenUserJS codebase. But have no clear idea where on supposed to look. If I want to add the Creative Commons license typed to the website.

[1]: https://openuserjs.org/scripts/Jeni4/Torrentz2_(dot)_eu_magnet_link
[2]: https://github.com/netravnen/torrentz2-magnet-link

**Comment by Martii** [comment]
> Can a/the reason be the 'CC BY-XYZ' licenses is not available fx here, in this dropdown ?

Use the correct SPDX code for the Content License of `CC-BY-SA 3.0`. The lack of a hyphen is an invalid SPDX code and also this is the only thing that is keeping your script from being deleted automatically during any sweeps but still maintain the legacy licensing that you agreed to from the TOS.  CC is also not OSI approved but properly dual licensing is recommended in your case.


And you must declare the Code License of `MIT` which is what the default TOS  is given permission by you or whatever you choose from the drop down list for dual licensing. Please ensure the Code License is the last `@license` or it will continue to reject.  Additional information has been posted for almost two months at https://openuserjs.org/user/add/scripts#user-block-license

> Would be nice to have it fixed.

Then you need to fix it...

e.g.

``` js
// @license CC-BY-SA 3.0; http://creativecommons.org/licenses/by-sa/3.0/
// @license MIT
```

... to allow editing/updating of scripts.

Just being clear here is all... You have two choices... fix the script or delete it. We have two choices... wait for you to decide and you continue to honor the TOS in it's legacy formor remove the script and possibly the account for TOS violation. Licensing must be fully declared.

> So amongst other the license in the info section is not displayed wrongly like this.

As you may have guessed by hopefully actually reading all the links by now... this is user error and not a server error. e.g. the ball is in your court to abide by the existing TOS that has been in place for years.

OUJS Staff.

**Comment by netravnen** [comment]
Makes sense! (My mistake. I now know.)

## Idea for minor improvement
The error 400 page message could (hopefully) point out fx:
_read [TOS section][1] 'license' for clarification of OSI models and find the list of OSI approved licenses [here][2]_

This was my biggest confusion when getting the error message. "What is meant by 'OSI primary'". Finding the link to [opensource.org][2] hidden in the TOS text was actually the biggest help understanding the issue on my end. As I would not normally expect to find the above mentioned info in a TOS declaration, unless told otherwise in an error message. (would usually look to an available  site FAQ for this kind of info.)

[1]: https://openuserjs.org/about/Terms-of-Service#licensing
[2]: https://opensource.org/licenses/category

**Comment by Martii** [comment]
> "What is meant by 'OSI primary'".

Dual licensing. Primary and secondary. Primary OSI `MIT` and secondary `CC-BY-SA-3.0`, etc.

> Finding the link to opensource.org hidden in the TOS text 

Interesting perception but not hidden... bright orange versus black in the TOS. Both sites are linked their for convenience nearly about the same time we started licensing enforcement since a lot of authors weren't abiding by it especially with CC and ND.

> would usually look to an available site FAQ for this information

Try the link I gave you [above](https://github.com/OpenUserJS/OpenUserJS.org/issues/1314#issuecomment-354565799):

> https://openuserjs.org/user/add/scripts#user-block-license

With the exception of the expanding/collapsing accordion e.g. you have to click it to expand it *(for a few weeks we expanded it for everyone but it's a link so one can read the comment in the above panel states clicking)*...  it's been there as long as myself with this project just outlined a little better and conservation of screen real estate is important especially if one already knows about it e.g. that is the FAQ document in actual context. Any commits that modify the blocks descriptions usually have the DOC label which includes the FAQs.

> 'license' for clarification of OSI models and find the list of OSI approved licenses [here](https://opensource.org/licenses/category)

Can't do that at your "here" because there is a finite number of accepted, current, non-expired, non-deprecated, and "compatible" *(might notice that from the error message)*.... CC's ND is not compatible for example and a definite TOS violation... so that's why the drop down exists with the accepted values on new scripts and existing scripts.

Secondly I still need to look up the RFC for status messages and ensure that HTML can or can not be allowed for HTTP/1.0, HTTP/ 1.1 and the upcoming HTTP/2.0 standard ... as well as hope that none of the external sites, like the GH webhook... will currently or ever care. So an issue. Additionally the GH webhook parsing dependency will not stay non-breaking long enough so it's currently frozen by me since I do the dep updates.

So in summary it's a good shared idea that I'm still pondering over however there are a lot of obstacles... including the holiday season. More changes will come down the line as I stated on OUJS itself.

**Comment by netravnen** [comment]
Hi Marti,

On 31 December 2017 at 02:09, Marti Martz <notifications@github.com> wrote:

> "What is meant by 'OSI primary'".
>
> Dual licensing. Primary and secondary. Primary OSI MIT and secondary
> CC-BY-SA-3.0, etc.
>
​Makes sense after you explain it.
Did not make sense when I read for the first time yesterday.​


> Finding the link to opensource.org hidden in the TOS text
>
> Interesting perception but not hidden... bright orange versus black in the
> TOS.​ (...)​
>
> ​Tend to forget I use f.lux <https://justgetflux.com/> to skew my color
pallet on the screen itself. (Drowns out the orange when I view the page in
the afternoon/nigth).​

>
> would usually look to an available site FAQ for this information
>
> Try the link I gave you above
> <https://github.com/OpenUserJS/OpenUserJS.org/issues/1314#issuecomment-354565799>
> :
>
​Understood the first time. But thx anyways for pointing out again. :)

> https://openuserjs.org/user/add/scripts#user-block-license
>
> ​(...)
> e.g. that is the FAQ document in actual context. Any commits that modify
> the blocks descriptions usually have the DOC label which includes the FAQs.
>
> ​Will look out for this. Thx.

> 'license' for clarification of OSI models and find the list of OSI
> approved licenses here <https://opensource.org/licenses/category>
>
> Can't do that at your "here" because there is a finite number of accepted,
> current, non-expired, non-deprecated, and "compatible" *(might notice
> that from the error message)*.... CC's ND is not compatible for example
> and a definite TOS violation... so that's why the drop down exists with the
> accepted values on new scripts and existing scripts.
>
​Reasonable.

> Secondly I still need to look up the RFC for status messages and ensure
> that HTML can or can not be allowed for HTTP/1.0, HTTP/ 1.1 and the
> upcoming HTTP/2.0 standard ... as well as hope that none of the external
> sites, like the GH webhook... will currently or ever care. So an issue.
> Additionally the GH webhook parsing dependency will not stay non-breaking
> long enough so it's currently frozen by me since I do the dep updates.
>
Hmm. ​If dropping the HTML part and just point to the OUJS TOS works. That
could be good enough.

> So in summary it's a good shared idea that I'm still pondering over
> however there are a lot of obstacles... including the holiday season.
>
> ​​ ​​​Ahh. Holidays. What a great time they give us. ^_^​

> More changes will come down the line as I stated on OUJS itself
>
> Looked in the forum.

Have a hard time finding the statement referred with no link. (unless you
mean https://openuserjs.org/announcements)

​/netravnen​

**Comment by Martii** [comment]
> If dropping the HTML part and just point to the OUJS TOS works

TOS is already linked on every page... if it were to be referenced with or without html it would point to the context FAQ. Usually most people read what's on the screen rather than extreme hand holding... but when dev duties resume I'll investigate the RFCs including new ones... takes time.

> Have a hard time finding the statement referred with no link

[Up there](https://github.com/OpenUserJS/OpenUserJS.org/issues/1314#issuecomment-354565799) ... repeated here of "[a discussion announcement](https://openuserjs.org/announcements/Licensing_enforcement)".

> unless you mean https://openuserjs.org/announcements)

More general but yes.

**Comment by Martii** [comment]
> Makes sense after you explain it.

Oh and primary is the defacto Userscript standard in metadata blocks e.g. FILO queue on all keys so that's typically out of scope for OUJS and more along the lines of GM standards and their peers... which is why it is briefly mentioned. e.g. one should know this when using a .user.js engine. :) I mention it a few times during the year and if you look at all the keys they state what is first *(primary)*, if multiple values, etc. `@license` can be multiple but last is primary because it's First In Last Out *(a.k.a Last Out is First In)*.

**Comment by netravnen** [comment]
​​
>
​
​
>​
​
If dropping the HTML part and just point to the OUJS TOS works
​
​
>
​
>
​
> TOS is already linked on every page... if it were to be referenced with
or without html it would point to the context FAQ. Usually most people read
what's on the screen rather than extreme hand holding... but when dev
duties resume I'll
​​
investigate the RFCs including new ones... takes time.

​"Usually most people read what's on the screen rather than extreme hand
holding" ... I've been a handful, Yes. I've now learned and been educated.

"
​
investigate the RFCs"... Come from a Network Engineer background. So mostly
a dilettante when hitting stuff ending in .js. (pun intended ~= joke)

​
​
>
​
> Have a hard time finding the statement referred with no link
​
​
>
​
>
> Up there ... repeated here of "a discussion announcement".

Ahh. Now I got what you meant.

​
​
>
​
> unless you mean https://openuserjs.org/announcements)
​
​
>
​
>
> More general but yes.

Noted.

**Comment by netravnen** [comment]
Thx A LOT(!) for helping this handful of 1 person out with understanding a
fair amount of information etc. concerning OUJS.

I have understood all relevant information for now.

Read the referenced material on the OUJS website.

My utmost thanks for your patience and time with me.

/netravnen

**Comment by joeytwiddle** [comment]
This still isn't clear enough, evidenced by my inability to publish a script today.

The 400 page (the one with the following message) does not offer any link to get help.

>  `@license` is not OSI primary and compatible in the metadata block(s).

Googling brought me to the [TOS page](https://github.com/OpenUserJS/OpenUserJS.org/wiki/Terms-of-Service). That linked me to the [OSI page](https://opensource.org/licenses/category) but the code I found on that page ('GPL') still didn't work!

I think what would help is a link from the 400 page and from the TOS/FAQ to a page with a list of accepted license codes.

Thanks!

**Comment by Martii** [comment]
> with a list of accepted license codes.

Doesn't quite work like that... put in pseudo mathematical terms it's https://github.com/shinnn/spdx-license-ids minus https://github.com/OpenUserJS/OpenUserJS.org/blob/master/controllers/blockSPDX.json . i.e. it's dynamic. These values could change if the dep is updated at any point or OSI deprecates a particular version *(like they did with older MPL)*... so it's not that difficult to go to https://openuserjs.org/user/add/scripts/new and see what new and existing scripts are allowed to have.

As you stated:

> I have eventually found the Author Tools which should help...

... is the time that it is needed for new users and perhaps some old ones. To allot any extra CPU time would be counter-productive and a definite hand holding imho. 

Jumping right into GH importing shouldn't be a new users first task. As I have already stated here the specs need to be confirmed when there is time and have very little for myself at the moment... this includes **from all** of those who actually want the possible feature on the error message page e.g. How about some comment help with some searches at the very minimum... not just me. 

Notifications are in ~5 places now... primary links are in two *(times two)* common places and additional on existing scripts... another one is a bit excessive imho but the help is welcomed with a validated/documented/referenced PR. e.g. I know how to put the equivalent of `innerHTML` in instead of `textContent` but what are the possible repercussions that may ensue if not standard. e.g. is the suggestion going to break importing esp. since the dep is currently frozen from updates due to breaking semver announcements.

Without these it's tabled in **my** duty lists... doesn't mean it's gone just not a very high priority in **my** tasks of coding/searching.

**Comment by netravnen** [comment]
Yes. We overlook the obvious.
Yes. Did not find the correct documentation in what in hindsight looks like obvious places on the website.
Yes. We were a bit to quick to just start asking questions.
Yes. We apologize.
And please do not take it (to much) to heart we kept pestengiring in what is now looking like longer than necessary.
